src/mds/Locker.cc

```diff
@@ -4794,12 +4829,113 @@ void Locker::scatter_tick()
 	       << *lock << " " << *lock->get_parent() << dendl;
       continue;
     }
-    if (now - lock->get_update_stamp() < g_conf()->mds_scatter_nudge_interval)
+
+    if (!mdr && now - lock->get_update_stamp() < g_conf()->mds_scatter_nudge_interval)
       break;
     updated_scatterlocks.pop_front();
-    scatter_nudge(lock, 0);
+
+    dout(20) << __func__ << " lock: " << *lock << " of " << *lock->get_parent()
+      << ", update_stamp: " << lock->get_update_stamp() << ", lock's parent's rstat_dirty_from:"
+      << static_cast<CInode*>(lock->get_parent())->get_rstat_dirty_from() << ", lock's parent's subdir rstat_dirty_from:"
+      << static_cast<CInode*>(lock->get_parent())->get_dirfrags_rstat_dirty_from()
+      << ", mdr's rstats_propagate_time: " << rstat_propagate_time << dendl;
+
+    if (lock->get_type() == CEPH_LOCK_INEST) {
+
+      CInode* ino = static_cast<CInode*>(lock->get_parent());
+      if (lock->is_nudging()) {
+        // We need to keep an eye on nestlocks that are of the current mdr's interest
+        // and are already nudged by scatter_tick or other rstat flush operation(Although
+        // there can be at most only one client rstat flush request at any time, there
+        // could be multiple slave requests at the same time if the current mds is auth
+        // for multiple disjoint subtrees).
+        if (mdr && ino->need_to_nudge(rstat_propagate_time)
+            && !lock->is_nudged_by(mdr->reqid)) {
+
+          dout(20) << __func__ << " " << mdr->reqid << " nestlock of " << *lock->get_parent()
+            << " already nudged, adding callback" << dendl;
+          lock->add_waiter(SimpleLock::WAIT_STABLE, new C_Locker_Nudge(this, mdr));
+          // nudged_by field is used to record mdrs that are interested in this lock
+          // and make sure that each mdr is recorded only once. Once an mdr is add to
+          // nudged_by field, it won't be add again.
+          lock->add_nudged_by(mdr->reqid);
+        }
+        continue;
+      }
+
+      // We need to keep an eye on all nestlocks that's needed to
+      // be nudged, because, when an rstat flush op comes, some
+      // nestlocks that's needed maybe not be in the updated_scatterlocks
+      // queue, as they may already be nudged by some previous scatter_tick
+      // invocation. So, we have to add all nudging nestlocks to "nudging_nestlocks",
+      // so that when rstat flush request comes, no interested nestlocks
+      // can be missed.
+      lock->set_nudging(nudging_nestlocks,
+          (mdr && ino->need_to_nudge(rstat_propagate_time))
+            ? mdr->reqid : metareqid_t());
```

Comment:
set 'nudging' before checking if nestlock is nudge-able? Is it too early?

Reply:
The nudging state's life time is from the point the scatter_nudge is invoked to that the nestlock finish waiting(unfreeze, single auth, or stable). the reason why we have to inclose these waiting interval is that the nestlock isn't in the updated_scatterlocks queue, so we have to trace them.

Reply:
how about
1. add these not-nudge-able lock to another list, and add a waiter context.
2. when waiter context is executed, move it back to updated_scatterlocks list.

with this approach, we don't need the 'nugding'  state

Reply:
I don't quite follow you. the nudge events are all independent and of different kinds(unfreeze, stable, single auth), how to let add a waiter for all those events? and what if some other rstat flush op arrives before the waiter is invoked and it has

Reply:
I don't quite follow you. the nudge events are all independent and of different kinds(unfreeze, stable, single auth), how to let add a waiter for all those events? and what if some other rstat flush op arrives before the waiter is invoked?  there could be locks in that queue that are of the new rstat flush's interest, too. how to make it wait for those locks?

Reply:
these events can't happen at the same time. just add waiter for individual event. 

rstat flusher just check the if any locks (it interests) is in the new list and wait.

Reply:
Oh, finally, I got your point:-)
Actually, I'm already doing exactly what you are saying:-)

The Locker::nudging_nestlocks is the list that you're talking about; the ScatterLock::set_nudging() method is to add the nestlock to the nudging_nestlocks list; ScatterLock::is_nudging() method is to test whether a nestlock is in the nudging_nestlocks list; ScatterLock::clear_nudging() method is to remove a nestlock from the nudging_nestlocks. 

The only difference is that I added a redundant NUDGING state, and this state seems really unnecessary now, all those methods can be easily modified to run without this state. I think I can simply remove the state:-)